### DISCUSSION TITLE
:tada: [Google Summer Code Program 2023] - Helper to upgrade the projects

### DISCUSSION THREAD

**Discussion by camilamacedo86** [kubernetes-sigs/kubebuilder#3283] [Announcements]
🔈 Kubebuilder is part of the [Google Summer Code Program 2023](https://summerofcode.withgoogle.com/). This project is applied under the [CNFC organization](https://github.com/cncf/mentoring/blob/main/summerofcode/2023.md#kubebuilder).

**Contributor Proposal Submission period: March 20 - April 4 (1800 UTC)**

> Starting Monday, March 20 at 1800 UTC, contributors will be able to submit their proposals via the program website. 

**To attend the program, you must be an open-source beginner or a student**. Check out the [GSoC FAQ entry for eligibility requirements](https://developers.google.com/open-source/gsoc/faq#what_are_the_eligibility_requirements_for_participation). Also, you can check out some [videos](https://developers.google.com/open-source/gsoc/videos) to know more about. 

You must apply as a student (GSoC Contributor) to the program and propose this project.  It does not need to be big or verbose. Then, you can be stunned by its [timeline](https://developers.google.com/open-source/gsoc/timeline) to know if your proposal was accepted and when you should begin to work on the project. We recommend you make a few small contributions against contributors to check how that works and see if that fits your purposes to apply.  

## Helper to upgrade the projects

Things change, and we constantly grow the [KubeBuilder](https://github.com/kubernetes-sigs/kubebuilder), providing new features and bug fixes. Also, sometimes it is required to address incompatible changes via new plugin versions. However, all changes and growth bring some complexities to its users keeping their solutions upgraded and adopting all that is new. 

The primary motivation of this project is to provide a helper via a command CLI that will automate a common and manual part of this process and try to make it less painful. Also, this project will add a lot of value for Kubebuilder, and its maintainers since it can encourage their users to move forward more frequently. We might be able to use this feature to create lovely automation using git and provide GitHub actions in the future. Note that we have a design proposal to develop the initial version of this feature, which is expected in this project. However, your ideas and input to solve this challenge will be very welcome!

See the proposal: https://github.com/kubernetes-sigs/kubebuilder/blob/master/designs/helper_to_upgrade_projects_by_rescaffolding.md

🚀 You can learn more about the project by looking at the issue: https://github.com/kubernetes-sigs/kubebuilder/issues/3244 

## How to do the proposal?

- Read the design doc, which defines how that should be made: https://github.com/kubernetes-sigs/kubebuilder/blob/master/designs/helper_to_upgrade_projects_by_rescaffolding.md
- Ensure that you did a few/couple of PRs in the project to see how it works

**Then, if what is not clear, the motivation and how that works:**
- a) Follow the QuickStart to see how you build a project with the tool: https://book.kubebuilder.io/quick-start.html (5 min)
- b) Read and check out how the users migrate from one version to another today before the helper. The goal is to automate the first part (to generate the project again based on the fino tracked in the PROJECT file):  (+ 5/10 mins ) https://book.kubebuilder.io/migration/migration_guide_gov3_to_gov4.html

**💡 Your proposal lets us know how you pretend to do the project. You do not need to be very lengthy**

###  💁‍♀️ What does Kubebuilder? How is it used?

Kubebuilder is often used as a CLI tool that generates projects and provides features to help people create "Operator Solutions".
"Operator solutions" are projects which follow the [Operator pattern](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/) and [extend the K8S API](https://kubernetes.io/docs/tasks/extend-kubernetes/) and has the purpose of managing solutions on K8s clusters.

TL'DR:

> All Operators are made of [controllers](https://github.com/kubernetes-sigs/controller-runtime), but not all sets of controllers are operators. Operators are sets of controllers that encode human operational knowledge for a specific application (e.g. the Prometheus Operator). Controllers are more general -- for instance, the namespace service account controller isn't an operator because it doesn't know a specific application -- it's very general. Kubebuilder helps constructs sets of controllers. Some of those are Operators, and some might not be.
> 
> Also, Kubebuilder can be used as a lib to help other projects extend their features and/or create nice [plugins](https://book.kubebuilder.io/plugins/plugins.html) which can be used as specific helpers. For example, [Operator-SDK](https://github.com/operator-framework/operator-sdk) uses Kubebuilder and scaffolds the same plugins/projects but also provide additional features and other plugins on top.

See:

- [Readme Overview](https://github.com/kubernetes-sigs/kubebuilder#kubebuilder)
- [Readme Motivatios](https://github.com/kubernetes-sigs/kubebuilder#motivation)
- [Who is this for](https://book.kubebuilder.io/introduction.html#who-is-this-for)

### 💡 Examples of its usage:

See https://operatorhub.io/ to have a better idea

### ❓ Why Operators?

By using operators, it’s possible to provide all expected resources and manage them dynamically, programmatically, and at execution time. To illustrate this idea, imagine if someone accidentally changed a configuration or removed a resource by mistake; in this case, the operator could fix it without human intervention.

What is the difference between Kubebuilder and SDK?
See: https://sdk.operatorframework.io/docs/faqs/#what-are-the-the-differences-between-kubebuilder-and-operator-sdk

### 🚀 How can I begin to become familiarized with the Project?

OpenSource projects usually have a Contribution guide; see [CONTRIBUTING](https://github.com/kubernetes-sigs/kubebuilder/blob/master/CONTRIBUTING.md) and also [VERSIONING](https://github.com/kubernetes-sigs/kubebuilder/blob/master/VERSIONING.md) for further details.

Also, feel free to check the open issues in the project and begin to work and contribute with anything you wish.
Your help and contributions 🥇 are very welcome.

The maintainers actively manage the issues list and try to highlight issues suitable for newcomers (You can filter by [good+first+issue](https://github.com/kubernetes-sigs/kubebuilder/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22) also contributions to help on the docs can fit well for the first ones), but you are free to help as please you. The project follows the typical GitHub pull request model. Before you start any work, please either comment on a current issue or file a new one.

### ✋ How to get help?

Feel free to ping us and join the #kubebuilder channel in the kubernetes.slack.com slack.
You also can open issues in its repo if you see that fits.

### 💡 How to skill up and get started?

- You can follow the Kubebuilder QuickStart and tutorial: https://book.kubebuilder.io/quick-start.html
- Check out [Deploy Mmage Plugin](https://book.kubebuilder.io/plugins/deploy-image-plugin-v1-alpha.html) developed in the latest Google Summer Program 2022 to understand how the projects are done. You will see that by using this plugin, you will have a project with the whole code to distribute and manage a solution on a Kubernetes cluster. You can also look at the demonstration [video](https://www.youtube.com/watch?v=UwPuRjjnMjY) done by the student in the Community meeting. 
- Another amazing project built last year as part of this program was the [Grafana plugin](https://book.kubebuilder.io/plugins/grafana-v1-alpha.html); feel free to check its [demonstration video](https://www.youtube.com/watch?v=-w_JjcV8jXc) as well 

###  How are the mentors of this project? 
@Kavinjsir , @varshaprasad96 , @camilamacedo86 
However, if you need to contact us, please create a group and add us to them so we can try to answer as soon as possible.

Students who reached out so far.
----

| Slack | Github |
| -- | -- |
| - | @xiao-jay |
| Sajiyah Salat | @Sajiyah-Salat |
| - | @octonawish-akcodes  |
| - | @yyy1000 |
| Ashutosh | @ashutosh887)

**Comment by camilamacedo86** [comment]
@Kavinjsir, would you like to add here your experience or any info that you think can help the potential GSoC Contributor to know how to apply and how to make their application?

**Reply by Kavinjsir** [reply]
Sure thing! @camilamacedo86 glad to have you ping me!

I think you've already provided great knowledge above 👍🏼 . That's actually where I got started last year 😃.
Since it is really an amazing experience from my side, I wish our contributors can enjoy the journey here.
In general, I would suggest:
1. Play with Kubebuilder to learn its functionality (read the Kb book if necessary, and follow the tutorial)
2. Get familiar with the community
    - Fork and learn the source code
    - Learn about the contribution rule and development practices
    - Kindly raise questions if you need help OR discussion
3. Be well-prepared for the GSoC project
    - Track the origin issue and try to understand the context of the idea
    - Learn from the designed docs (proposal), if existing

**Comment by ashutosh887** [comment]
Hi
Even I'm interested to be a part of this Project

Please add me to the Students who've reached out so far/let me know how to reach out to you regarding this 🙂

**Comment by Sajiyah-Salat** [comment]
As a contributor to kubebuilder I have shared my experience in the form of blog and video. Check it out [video](https://www.youtube.com/watch?v=y7TD612A5T4)
blog will be out soon!

**Comment by ashutosh887** [comment]
@camilamacedo86 I'd also reached out to you for GSOC 😄

**Comment by camilamacedo86** [comment]
**Contributor Proposal Submission period: March 20 - April 4 (1800 UTC)**

> Starting Monday, March 20 at 1800 UTC, contributors will be able to submit their proposals via the program website. Each contributor may submit up to 3 proposals to the program. They can submit multiple proposals to the same organization, or spread their proposals across multiple organizations.

**Reply by ashutosh887** [reply]
Hi @camilamacedo86 
Is there any specific template that needs to be followed for proposal?

**Reply by Sajiyah-Salat** [reply]
No, there is no specific template. Take inspiration from gsoc common templates.

**Comment by camilamacedo86** [comment]
Hi @xiao-jay , @Sajiyah-Salat, @octonawish-akcodes, @yyy1000, @ashutosh887.

We did not receive your proposals on the platform https://summerofcode.withgoogle.com/ yet. Please, ensure that you add your proposal to their prior April 4 (1800 UTC). Thank you, and good luck !!!
We are looking forward to working with you.

**Comment by Sajiyah-Salat** [comment]
Hey camilla, can you please provide me a link where proposal  need to be added.

**Reply by xiao-jay** [reply]
just subbmit to cncf is ok https://summerofcode.withgoogle.com/programs/2023/organizations/cncf/create-proposal

**Reply by ashutosh887** [reply]
Yes @Sajiyah-Salat 
First create a contributor account on the platform, fill the required details and then proceed.

**Reply by Sajiyah-Salat** [reply]
Ohk, thanks

**Comment by sayantani11** [comment]
Hey I am creating a proposal for the same :)

**Comment by ashutosh887** [comment]
Congratulations to all those who got accepted for GSOC 🚀✨

**Comment by camilamacedo86** [comment]
We received a multitude of applications and, while each one was exceptional, we were tasked with the difficult decision of selecting just one. The votes pointed us towards @yyy1000, and we are eager to see their contributions to our project.

Our heartfelt thanks to every participant. 🥇 Your entries were beyond amazing, which made the selection process incredibly challenging. Although we could select only one student this time, this shouldn't dampen your spirits.

We strongly encourage each one of you to continue honing your skills and chasing your dreams. You're always welcome to contribute to the project and utilize Kubebuilder as a tool for independent learning, regardless of your involvement with this Program. We are here to support you at every step. Please, feel free to propose PRs and work on areas that you see fit. 

🎉 We're also excited to announce that despite not being under the official Google Summer Code Program, @yyy1000 has chosen to remain committed to this project. We will adhere to the same timelines and guidelines as we would have under the Google Summer Code Program.

**Remember, opportunities are always around the corner, so never stop reaching out for them!**

@varshaprasad96 @Kavinjsir

**Comment by yyy1000** [comment]
To be honest, I was very sad when I found out I had been rejected from GSoC. But you know what, failure is always a part of life. Therefore, I decided to  continue this project. I will do my best to complete this project and make more contributions to make kubebuilder better.

**Reply by Sajiyah-Salat** [reply]
Congrats @yyy1000 for being a selected participant. I wish you best of luck

**Comment by ashutosh887** [comment]
Congratulations to @yyy1000 💯

@camilamacedo86 
Even I'm willing to contribute to the Project
I'll be happy to do my contribution to improve KubeBuilder ✨

**Comment by camilamacedo86** [comment]
A record has been created to showcase the outcomes of this project. You can view it here: https://youtu.be/7997RIbx8kw?si=sWNaxddIFQTILVZA

🎉 I'd like to extend my heartfelt congratulations to the student responsible for this achievement @yyy1000 . Special thanks to @varshaprasad96  and @Kavinjsir for their invaluable mentorship and guidance throughout the project.

**Reply by Sajiyah-Salat** [reply]
@yyy1000 Congrats on completing th project. you have done a great job.

**Reply by ashutosh887** [reply]
Great work @yyy1000

**Comment by camilamacedo86** [comment]
Hi @yyy1000,

On September 6, 2023, you shared a demo video showcasing your contributions, sent via Slack (private message) to me, @Kavinjsir , and @varshaprasad96 . At that time, you authorised us to share it on the Kubebuilder YouTube channel, documentation, and any other relevant channels or venues—without requiring disclaimers or additional conditions.

Just registering this here for transparency.
I’d appreciate it if you could confirm publicly here as well.

Thanks!

**Comment by yyy1000** [comment]
Hey @camilamacedo86,
Sure, feel free to share it publicly and I'm very happy my work could be useful as well. : )
